src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+    that is {e already rejected} with [exn]. *)
 
-val fail : exn -> 'a t
-  (** [fail e] is a thread that fails with the exception [e]. *)
 
-val fail_with : string -> 'a t
-  (** [fail_with msg] is a thread that fails with the exception
-      [Failure msg]. *)
 
-val fail_invalid_arg : string -> 'a t
-  (** [fail_invalid_arg msg] is a thread that fails with the exception
-      [Invalid_argument msg]. *)
+(** {3 Callbacks} *)
 
 val bind : 'a t -> ('a -> 'b t) -> 'b t
-  (** [bind t f] is a thread which first waits for the thread [t] to
-      terminate and then, if the thread succeeds, behaves as the
-      application of function [f] to the return value of [t].  If the
-      thread [t] fails, [bind t f] also fails, with the same
+(** [Lwt.bind p_1 f] makes it so that [f] will run when [p_1] is {{: #TYPEt}
+    {e fulfilled}}.
+
+    The value that [p_1] is fulfilled with is passed to [f]. [f] then returns
+    another promise, [p_2]. This second promise means [f] can spawn more I/O
+    operations – they will create [p_2].
+
+    [Lwt.bind] itself returns immediately. It only attaches the callback [f] to
+    [p_1] – it does not wait for [p_2]. {e What} [Lwt.bind] returns is yet a
+    third promise, [p_3]. Roughly speaking, fulfillment of [p_3] represents both
+    [p_1] and [p_2] becoming fulfilled, one after the other.
+
+    A minimal example of this is an echo program:
+
+{[
+let () =
+  let p_3 =
+    Lwt.bind
+      Lwt_io.(read_line stdin)
+      (fun line -> Lwt_io.printl line)
+  in
+  Lwt_main.run p_3
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    Rejection of [p_1] and [p_2], and raising an exception in [f], are all
+    forwarded to rejection of [p_3].
+
+    The rest is (1) a precise description, and (2) notes on why [Lwt.bind] is
+    always written using syntactic sugar.
+
+    {b (1)} [Lwt.bind] returns a promise [p_3] immediately. [p_3] starts out
+    pending, and is resolved as follows:
+
+    - The first condition to wait for is that [p_1] becomes resolved. It does
+      not matter whether [p_1] is already resolved when [Lwt.bind] is called, or
+      becomes resolved later – the rest of the behavior is the same.
+    - If and when [p_1] becomes resolved, it will, by definition, be either
+      fulfilled or rejected. Fulfilling is the interesting case, but rejection
+      is handled very simply, so we cover rejection of [p_1] first.
+    - If [p_1] is rejected, [p_3] is rejected with the same exception.
+    - If [p_1] is fulfilled, with value [v], [f] is applied to [v].
+    - [f] may finish by returning the promise [p_2], or raising an exception.
+      Again, the exception case is simpler, so we cover that first.
+    - If [f] raises an exception, [p_3] is rejected with that exception.
+    - Finally, the remaining case is when [f] returns [p_2]. From that point on,
+      [p_3] is effectively made into a reference to [p_2]. This means they have
```

Comment:
The notion of references is never discussed before. I don't think it's an issue, but then I'm familiar with the library. How confusing is it for a beginner?

Reply:
I'm not sure what to do here, and I'm tempted to push it to a future rewrite (I'll give this page at least one more thorough look as the manual comes together, plus anyone else is welcome to).

The "references" I'm mentioning here are an extra-library concept, as in runtime-level references (the same way `Lwt.t`s and their corresponding `Lwt.u`s are references to the same object in memory). Not `ref` references. I'm hoping that enough readers of these docs will be familiar with this idea. Of course, I may be wrong.

Reply:
I guess it's not that beginner friendly. We could try wording like "made to behave exactly the same as `p_2`," but I want to convey a stronger image that `p_3` *is* `p_2`, because it pretty much is, given how Lwt is implemented :) "Behave exactly the same as `p_2`" leaves me with some kind of uncertainty about the extent to which that's true, etc.